x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/MlAutoscalingResourceTracker.java

```diff
@@ -0,0 +1,160 @@
+            .clear()
+            .setOs(true)
+            .setTimeout(timeout)
+            .execute(
+                ActionListener.wrap(
+                    nodesStatsResponse -> listener.onResponse(
+                        nodesStatsResponse.getNodes()
+                            .stream()
+                            .collect(Collectors.toMap(nodeStats -> nodeStats.getNode().getId(), NodeStats::getOs))
+                    ),
+                    listener::onFailure
+                )
+            );
+    }
+
+    static void getMemoryAndCpu(
+        ClusterState clusterState,
+        MlMemoryTracker mlMemoryTracker,
+        Map<String, OsStats> osStatsPerNode,
+        ActionListener<MlAutoscalingStats> listener
+    ) {
+        Set<String> nodesWithRunningJobs = new HashSet<>();
+        long memoryBytesSum = osStatsPerNode.values()
+            .stream()
+            .map(s -> s.getMem().getAdjustedTotal().getBytes())
+            .collect(Collectors.summingLong(Long::longValue));
+
+        long modelMemoryBytesSum = 0;
+        long extraSingleNodeModelMemoryInBytes = 0;
+        int extraSingleNodeProcessors = 0;
+        int minNodes = 0;
+        int extraProcessors = 0;
+
+        final MlAutoscalingContext autoscalingContext = new MlAutoscalingContext(clusterState);
+
+        for (var task : autoscalingContext.anomalyDetectionTasks) {
+            String jobId = ((OpenJobAction.JobParams) task.getParams()).getJobId();
+            if (AWAITING_LAZY_ASSIGNMENT.equals(task.getAssignment())) {
+                extraSingleNodeModelMemoryInBytes = Math.max(
+                    extraSingleNodeModelMemoryInBytes,
+                    mlMemoryTracker.getAnomalyDetectorJobMemoryRequirement(jobId)
+                );
+                extraSingleNodeProcessors = Math.max(extraSingleNodeProcessors, 1);
+                ++extraProcessors;
+            } else {
+                modelMemoryBytesSum += mlMemoryTracker.getAnomalyDetectorJobMemoryRequirement(jobId);
+                nodesWithRunningJobs.add(task.getExecutorNode());
+                minNodes = 1;
+            }
+        }
+        for (var modelAssignment : autoscalingContext.modelAssignments.entrySet()) {
+            if (AssignmentState.STARTING.equals(modelAssignment.getValue().getAssignmentState())
+                && modelAssignment.getValue().getNodeRoutingTable().isEmpty()) {
+                extraSingleNodeModelMemoryInBytes = Math.max(
+                    extraSingleNodeModelMemoryInBytes,
+                    modelAssignment.getValue().getTaskParams().estimateMemoryUsageBytes()
+                );
+                final int unassignedProcessors = modelAssignment.getValue().getTaskParams().getNumberOfAllocations() * modelAssignment
+                    .getValue()
+                    .getTaskParams()
+                    .getThreadsPerAllocation();
+
+                extraSingleNodeProcessors = Math.max(extraSingleNodeProcessors, unassignedProcessors);
```

Comment:
We don't need every allocation to be on a single node - in theory each allocation could be put on a different node. So the extra processors required on a single node should only be the threads per allocation.